File: MATLAB/aceg/aceg_O_AVP.m
% ---------------------------------------------------------------------
% Package:    TopiCo (https://github.com/AIS-Bonn/TopiCo)
% Version:    2021-03-18 12:09:55
% Maintainer: Marius Beul (mbeul@ais.uni-bonn.de)
% License:    BSD
% ---------------------------------------------------------------------

% Software License Agreement (BSD License)
% Copyright (c) 2021, Computer Science Institute VI, University of Bonn
% All rights reserved.
% Redistribution and use in source and binary forms, with or without
% modification, are permitted provided that the following conditions
% are met:
% 
% * Redistributions of source code must retain the above copyright
%   notice, this list of conditions and the following disclaimer.
% * Redistributions in binary form must reproduce the above
%   copyright notice, this list of conditions and the following
%   disclaimer in the documentation and/or other materials provided
%   with the distribution.
% * Neither the name of University of Bonn, Computer Science Institute
%   VI nor the names of its contributors may be used to endorse or
%   promote products derived from this software without specific
%   prior written permission.
% 
% THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
% "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
% LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
% FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
% COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
% INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
% BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
% LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
% CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
% LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
% ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
% POSSIBILITY OF SUCH DAMAGE.
%% --------------------------------------------------------------------

function [t] = aceg_O_AVP(P_init,V_init,A_init,P_wayp,V_wayp,A_wayp,~,~,~,~,J_max,J_min) %#codegen
% Generated on 28-Aug-2019 12:21:18
coder.inline('default');

l2 = A_init.^2;
l3 = A_init.^3;
l5 = A_wayp.^2;
l6 = A_wayp.^3;
l8 = J_min.^2;
l9 = J_min.^3;
l11 = J_max.^2;
l12 = J_max.^3;
l13 = V_init.^2;
l14 = V_wayp.^2;
l17 = sqrt(3.0);
l18 = sqrt(6.0);
l30 = A_init.*J_min.*J_max.*V_init.*2.4e+1;
l31 = A_wayp.*J_min.*J_max.*V_wayp.*2.4e+1;
l4 = l2.^2;
l7 = l5.^2;
l10 = l8.^2;
l20 = J_min.*l12.*2.0;
l21 = J_max.*l9.*4.0;
l22 = J_min.*l3.*8.0;
l23 = J_max.*l3.*8.0;
l24 = J_min.*l6.*8.0;
l25 = J_max.*l6.*8.0;
l26 = P_init.*l12.*2.4e+1;
l27 = P_wayp.*l12.*2.4e+1;
l28 = V_init.*l12.*1.2e+1;
l29 = V_wayp.*l12.*1.2e+1;
l35 = -l30;
l36 = J_min.*J_max.*l2.*1.2e+1;
l37 = J_min.*J_max.*l5.*1.2e+1;
l38 = J_max.*V_init.*l2.*1.2e+1;
l39 = A_init.*V_init.*l11.*2.4e+1;
l40 = J_max.*V_wayp.*l2.*1.2e+1;
l41 = J_max.*V_init.*l5.*1.2e+1;
l42 = J_max.*V_wayp.*l5.*1.2e+1;
l43 = A_wayp.*V_wayp.*l11.*2.4e+1;
l44 = l2.*l5.*6.0;
l45 = J_min.*P_init.*l11.*2.4e+1;
l46 = J_min.*P_wayp.*l11.*2.4e+1;
l47 = J_max.*V_init.*l8.*1.2e+1;
l48 = J_min.*V_init.*l11.*2.4e+1;
l49 = J_max.*V_wayp.*l8.*1.2e+1;
l50 = J_min.*V_wayp.*l11.*2.4e+1;
l51 = l2.*l8.*6.0;
l52 = l2.*l11.*6.0;
l53 = l5.*l8.*6.0;
l54 = l5.*l11.*6.0;
l55 = l8.*l11.*5.0;
l56 = V_init.*V_wayp.*l11.*2.4e+1;
l70 = l11.*l13.*1.2e+1;
l71 = l11.*l14.*1.2e+1;
l15 = l4.*3.0;
l16 = l7.*3.0;
l32 = -l23;
l33 = -l24;
l34 = -l26;
l57 = -l38;
l58 = -l42;
l59 = -l43;
l60 = -l44;
l61 = -l46;
l62 = -l48;
l63 = -l50;
l64 = -l51;
l65 = -l52;
l66 = -l53;
l67 = -l54;
l69 = -l56;
l73 = -1.0./(l10-l20-l21+l55);
l74 = 1.0./(l10-l20-l21+l55).^2;
l76 = l74.^2;
l78 = l15+l16+l40+l41+l57+l58+l60+l69+l70+l71;
l81 = l22+l25+l27+l31+l32+l33+l34+l35+l39+l45+l59+l61;
l84 = l28+l29+l36+l37+l47+l49+l62+l63+l64+l65+l66+l67;
l79 = l78.^2;
l80 = l78.^3;
l82 = l81.^2;
l85 = l84.^2;
l86 = l84.^3;
l88 = (l78.*-4.0)./(l10-l20-l21+l55);
l89 = (l78.*-1.2e+1)./(l10-l20-l21+l55);
l101 = l74.*l78.*l84.*(4.0./3.0);
l102 = l74.*l78.*l84.*7.2e+1;
l83 = l82.^2;
l87 = l85.^2;
l91 = l80.*1.0./(l10-l20-l21+l55).^3.*2.56e+2;
l92 = l74.*l82.*2.7e+1;
l94 = (l74.*l82)./2.0;
l95 = l74.*l85;
l97 = l86.*1.0./(l10-l20-l21+l55).^3.*2.0;
l100 = (l86.*1.0./(l10-l20-l21+l55).^3)./2.7e+1;
l105 = l76.*l79.*l85.*1.28e+2;
l107 = l82.*l86.*1.0./(l10-l20-l21+l55).^5.*4.0;
l108 = l76.*l78.*l82.*l84.*1.44e+2;
l93 = l76.*l83.*2.7e+1;
l98 = l95./3.0;
l104 = l78.*l87.*1.0./(l10-l20-l21+l55).^5.*1.6e+1;
l109 = l91+l93+l104+l105+l107+l108;
l110 = sqrt(complex(l109));
l111 = l17.*l110.*3.0;
l112 = (l17.*l110)./1.8e+1;
l113 = l92+l97+l102+l111;
l115 = l94+l100+l101+l112;
l114 = sqrt(complex(l113));
l116 = l115.^(1.0./3.0);
l118 = 1.0./l115.^(1.0./6.0);
l117 = l116.^2;
l121 = (l18.*l81.*l114.*-3.0)./(l10-l20-l21+l55);
l122 = (l18.*l81.*l114.*3.0)./(l10-l20-l21+l55);
l123 = (l84.*l116.*-2.0)./(l10-l20-l21+l55);
l124 = (l84.*l116.*-6.0)./(l10-l20-l21+l55);
l119 = l117.*3.0;
l120 = l117.*9.0;
l125 = l89+l95+l120+l124;
l126 = l88+l98+l119+l123;
l127 = sqrt(complex(l126));
l128 = 1.0./l125.^(1.0./4.0);
l130 = (l17.*l78.*l127.*1.2e+1)./(l10-l20-l21+l55);
l132 = (l17.*l73.*l85.*l127)./(l10-l20-l21+l55);
l134 = l17.*l117.*l127.*-9.0;
l135 = (l17.*l118.*l127)./6.0;
l137 = (l17.*l84.*l116.*l127.*-1.2e+1)./(l10-l20-l21+l55);
l136 = -l135;
l138 = l121+l130+l132+l134+l137;
l139 = l122+l130+l132+l134+l137;
l140 = sqrt(complex(l138));
l141 = sqrt(complex(l139));
l142 = (l118.*l128.*l140)./6.0;
l143 = (l118.*l128.*l141)./6.0;
t3 = [0.0;l135-l143;l135+l143;l136-l142;l136+l142];

l2 = t3.^2;
t7 = (J_max.*V_init.*2.0-J_max.*V_wayp.*2.0+J_min.^2.*l2-A_init.^2+A_wayp.^2-A_wayp.*J_min.*t3.*2.0+A_wayp.*J_max.*t3.*2.0-J_min.*J_max.*l2)./(J_max.^2.*t3.*2.0-J_min.*J_max.*t3.*2.0);

t1 = -(A_init-A_wayp+J_min.*t3+J_max.*t7)./J_max;

t6 = [0.0;0.0;0.0;0.0;0.0];

t5 = [0.0;0.0;0.0;0.0;0.0];

t4 = [0.0;0.0;0.0;0.0;0.0];

t2 = [0.0;0.0;0.0;0.0;0.0];

t = [t1, t2, t3, t4, t5, t6, t7];

end

File: ROS/topico/src/aceg_O_AVP.cpp
//
// Academic License - for use in teaching, academic research, and meeting
// course requirements at degree granting institutions only.  Not for
// government, commercial, or other organizational use.
//
// aceg_O_AVP.cpp
//
// Code generation for function 'aceg_O_AVP'
//

// Include files
#include "aceg_O_AVP.h"
#include "power.h"
#include "rt_nonfinite.h"
#include "sqrt.h"
#include "topico_wrapper_data.h"
#include "topico_wrapper_rtwutil.h"
#include <cmath>

// Function Definitions
void aceg_O_AVP(double P_init, double V_init, double A_init, double P_wayp,
                double V_wayp, double A_wayp, double J_max, double J_min,
                creal_T t[35])
{
  creal_T l2[5];
  creal_T t3[5];
  creal_T b_l78;
  creal_T l110;
  creal_T l113;
  creal_T l115;
  creal_T l116;
  double a_tmp;
  double ar;
  double b_l78_tmp;
  double im;
  double l118_im;
  double l118_re;
  double l11_tmp;
  double l12;
  double l130_im;
  double l130_re;
  double l132_re;
  double l135_re;
  double l2_tmp;
  double l3;
  double l5_tmp;
  double l6;
  double l74;
  double l76;
  double l78;
  double l78_tmp;
  double l81;
  double l81_tmp;
  double l84;
  double l84_tmp;
  double l85;
  double l86;
  double l8_tmp;
  double l95;
  double re;
  int k;
  //  ---------------------------------------------------------------------
  //  Package:    TopiCo (https://github.com/AIS-Bonn/TopiCo)
  //  Version:    2021-03-18 12:09:55
  //  Maintainer: Marius Beul (mbeul@ais.uni-bonn.de)
  //  License:    BSD
  //  ---------------------------------------------------------------------
  //  Software License Agreement (BSD License)
  //  Copyright (c) 2021, Computer Science Institute VI, University of Bonn
  //  All rights reserved.
  //  Redistribution and use in source and binary forms, with or without
  //  modification, are permitted provided that the following conditions
  //  are met:
  //
  //  * Redistributions of source code must retain the above copyright
  //    notice, this list of conditions and the following disclaimer.
  //  * Redistributions in binary form must reproduce the above
  //    copyright notice, this list of conditions and the following
  //    disclaimer in the documentation and/or other materials provided
  //    with the distribution.
  //  * Neither the name of University of Bonn, Computer Science Institute
  //    VI nor the names of its contributors may be used to endorse or
  //    promote products derived from this software without specific
  //    prior written permission.
  //
  //  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
  //  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
  //  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
  //  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
  //  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
  //  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
  //  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
  //  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
  //  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
  //  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
  //  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
  //  POSSIBILITY OF SUCH DAMAGE.
  //  --------------------------------------------------------------------
  //  Generated on 28-Aug-2019 12:21:18
  l2_tmp = A_init * A_init;
  l3 = rt_powd_snf(A_init, 3.0);
  l5_tmp = A_wayp * A_wayp;
  l6 = rt_powd_snf(A_wayp, 3.0);
  l8_tmp = J_min * J_min;
  l11_tmp = J_max * J_max;
  l12 = rt_powd_snf(J_max, 3.0);
  a_tmp = ((l8_tmp * l8_tmp - J_min * l12 * 2.0) -
           J_max * rt_powd_snf(J_min, 3.0) * 4.0) +
          l8_tmp * l11_tmp * 5.0;
  l74 = 1.0 / (a_tmp * a_tmp);
  l76 = l74 * l74;
  l78_tmp = J_max * V_init;
  b_l78_tmp = J_max * V_wayp;
  l78 = ((((((((l2_tmp * l2_tmp * 3.0 + l5_tmp * l5_tmp * 3.0) +
               b_l78_tmp * l2_tmp * 12.0) +
              l78_tmp * l5_tmp * 12.0) +
             -(l78_tmp * l2_tmp * 12.0)) +
            -(b_l78_tmp * l5_tmp * 12.0)) +
           -(l2_tmp * l5_tmp * 6.0)) +
          -(V_init * V_wayp * l11_tmp * 24.0)) +
         l11_tmp * (V_init * V_init) * 12.0) +
        l11_tmp * (V_wayp * V_wayp) * 12.0;
  l81_tmp = A_wayp * J_min;
  l81 = ((((((((((J_min * l3 * 8.0 + J_max * l6 * 8.0) + P_wayp * l12 * 24.0) +
                l81_tmp * J_max * V_wayp * 24.0) +
               -(J_max * l3 * 8.0)) +
              -(J_min * l6 * 8.0)) +
             -(P_init * l12 * 24.0)) +
            -(A_init * J_min * J_max * V_init * 24.0)) +
           A_init * V_init * l11_tmp * 24.0) +
          J_min * P_init * l11_tmp * 24.0) +
         -(A_wayp * V_wayp * l11_tmp * 24.0)) +
        -(J_min * P_wayp * l11_tmp * 24.0);
  l84_tmp = J_min * J_max;
  l84 = ((((((((((V_init * l12 * 12.0 + V_wayp * l12 * 12.0) +
                 l84_tmp * l2_tmp * 12.0) +
                l84_tmp * l5_tmp * 12.0) +
               l78_tmp * l8_tmp * 12.0) +
              b_l78_tmp * l8_tmp * 12.0) +
             -(J_min * V_init * l11_tmp * 24.0)) +
            -(J_min * V_wayp * l11_tmp * 24.0)) +
           -(l2_tmp * l8_tmp * 6.0)) +
          -(l2_tmp * l11_tmp * 6.0)) +
         -(l5_tmp * l8_tmp * 6.0)) +
        -(l5_tmp * l11_tmp * 6.0);
  l6 = l81 * l81;
  l85 = l84 * l84;
  l86 = rt_powd_snf(l84, 3.0);
  l95 = l74 * l85;
  l3 = rt_powd_snf(a_tmp, 5.0);
  l135_re = rt_powd_snf(a_tmp, 3.0);
  l110.re =
      ((((rt_powd_snf(l78, 3.0) / l135_re * 256.0 + l76 * (l6 * l6) * 27.0) +
         l78 * (l85 * l85) / l3 * 16.0) +
        l76 * (l78 * l78) * l85 * 128.0) +
       l6 * l86 / l3 * 4.0) +
      l76 * l78 * l6 * l84 * 144.0;
  l110.im = 0.0;
  coder::internal::scalar::b_sqrt(&l110);
  l132_re = 1.7320508075688772 * l110.re;
  l76 = 1.7320508075688772 * l110.im;
  if (l76 == 0.0) {
    re = l132_re / 18.0;
    im = 0.0;
  } else if (l132_re == 0.0) {
    re = 0.0;
    im = l76 / 18.0;
  } else {
    re = l132_re / 18.0;
    im = l76 / 18.0;
  }
  l12 = l74 * l6;
  l6 = l86 / l135_re;
  l3 = l74 * l78 * l84;
  l115.re = ((l12 / 2.0 + l6 / 27.0) + l3 * 1.3333333333333333) + re;
  l115.im = im;
  l113.re = ((l12 * 27.0 + l6 * 2.0) + l3 * 72.0) + l132_re * 3.0;
  l113.im = l76 * 3.0;
  coder::internal::scalar::b_sqrt(&l113);
  l116 = coder::power(l115);
  b_l78 = coder::b_power(l115);
  if (b_l78.im == 0.0) {
    l118_re = 1.0 / b_l78.re;
    l118_im = 0.0;
  } else if (b_l78.re == 0.0) {
    l118_re = 0.0;
    l118_im = -(1.0 / b_l78.im);
  } else {
    l86 = std::abs(b_l78.re);
    l76 = std::abs(b_l78.im);
    if (l86 > l76) {
      l76 = b_l78.im / b_l78.re;
      l3 = b_l78.re + l76 * b_l78.im;
      l118_re = (l76 * 0.0 + 1.0) / l3;
      l118_im = (0.0 - l76) / l3;
    } else if (l76 == l86) {
      if (b_l78.re > 0.0) {
        l76 = 0.5;
      } else {
        l76 = -0.5;
      }
      if (b_l78.im > 0.0) {
        l3 = 0.5;
      } else {
        l3 = -0.5;
      }
      l118_re = (l76 + 0.0 * l3) / l86;
      l118_im = (0.0 * l76 - l3) / l86;
    } else {
      l76 = b_l78.re / b_l78.im;
      l3 = b_l78.im + l76 * b_l78.re;
      l118_re = l76 / l3;
      l118_im = (l76 * 0.0 - 1.0) / l3;
    }
  }
  l110.re = l116.re * l116.re - l116.im * l116.im;
  l3 = l116.re * l116.im;
  l110.im = l3 + l3;
  l12 = l84 * l116.re;
  l76 = l84 * l116.im;
  ar = l12 * -2.0;
  im = l76 * -2.0;
  if (im == 0.0) {
    l6 = ar / a_tmp;
    l3 = 0.0;
  } else if (ar == 0.0) {
    l6 = 0.0;
    l3 = im / a_tmp;
  } else {
    l6 = ar / a_tmp;
    l3 = im / a_tmp;
  }
  l115.re = ((l78 * -4.0 / a_tmp + l95 / 3.0) + l110.re * 3.0) + l6;
  l115.im = l110.im * 3.0 + l3;
  coder::internal::scalar::b_sqrt(&l115);
  ar = l12 * -6.0;
  im = l76 * -6.0;
  if (im == 0.0) {
    l6 = ar / a_tmp;
    l3 = 0.0;
  } else if (ar == 0.0) {
    l6 = 0.0;
    l3 = im / a_tmp;
  } else {
    l6 = ar / a_tmp;
    l3 = im / a_tmp;
  }
  b_l78.re = ((l78 * -12.0 / a_tmp + l95) + l110.re * 9.0) + l6;
  b_l78.im = l110.im * 9.0 + l3;
  b_l78 = coder::c_power(b_l78);
  if (b_l78.im == 0.0) {
    l12 = 1.0 / b_l78.re;
    l6 = 0.0;
  } else if (b_l78.re == 0.0) {
    l12 = 0.0;
    l6 = -(1.0 / b_l78.im);
  } else {
    l86 = std::abs(b_l78.re);
    l76 = std::abs(b_l78.im);
    if (l86 > l76) {
      l76 = b_l78.im / b_l78.re;
      l3 = b_l78.re + l76 * b_l78.im;
      l12 = (l76 * 0.0 + 1.0) / l3;
      l6 = (0.0 - l76) / l3;
    } else if (l76 == l86) {
      if (b_l78.re > 0.0) {
        l76 = 0.5;
      } else {
        l76 = -0.5;
      }
      if (b_l78.im > 0.0) {
        l3 = 0.5;
      } else {
        l3 = -0.5;
      }
      l12 = (l76 + 0.0 * l3) / l86;
      l6 = (0.0 * l76 - l3) / l86;
    } else {
      l76 = b_l78.re / b_l78.im;
      l3 = b_l78.im + l76 * b_l78.re;
      l12 = l76 / l3;
      l6 = (l76 * 0.0 - 1.0) / l3;
    }
  }
  ar = 1.7320508075688772 * l78 * l115.re * 12.0;
  im = 1.7320508075688772 * l78 * l115.im * 12.0;
  if (im == 0.0) {
    l130_re = ar / a_tmp;
    l130_im = 0.0;
  } else if (ar == 0.0) {
    l130_re = 0.0;
    l130_im = im / a_tmp;
  } else {
    l130_re = ar / a_tmp;
    l130_im = im / a_tmp;
  }
  l132_re = 1.7320508075688772 * (-1.0 / a_tmp) * l85;
  ar = l132_re * l115.re;
  im = l132_re * l115.im;
  if (im == 0.0) {
    l132_re = ar / a_tmp;
    l95 = 0.0;
  } else if (ar == 0.0) {
    l132_re = 0.0;
    l95 = im / a_tmp;
  } else {
    l132_re = ar / a_tmp;
    l95 = im / a_tmp;
  }
  re = 1.7320508075688772 * l110.re;
  im = 1.7320508075688772 * l110.im;
  l85 = -9.0 * (re * l115.re - im * l115.im);
  l78 = -9.0 * (re * l115.im + im * l115.re);
  re = 1.7320508075688772 * l118_re;
  im = 1.7320508075688772 * l118_im;
  l3 = re * l115.re - im * l115.im;
  im = re * l115.im + im * l115.re;
  if (im == 0.0) {
    l135_re = l3 / 6.0;
    l74 = 0.0;
  } else if (l3 == 0.0) {
    l135_re = 0.0;
    l74 = im / 6.0;
  } else {
    l135_re = l3 / 6.0;
    l74 = im / 6.0;
  }
  re = 1.7320508075688772 * l84 * l116.re;
  im = 1.7320508075688772 * l84 * l116.im;
  ar = (re * l115.re - im * l115.im) * -12.0;
  im = (re * l115.im + im * l115.re) * -12.0;
  if (im == 0.0) {
    l110.re = ar / a_tmp;
    l110.im = 0.0;
  } else if (ar == 0.0) {
    l110.re = 0.0;
    l110.im = im / a_tmp;
  } else {
    l110.re = ar / a_tmp;
    l110.im = im / a_tmp;
  }
  l86 = 2.4494897427831779 * l81 * l113.re;
  l84 = 2.4494897427831779 * l81 * l113.im;
  ar = l86 * -3.0;
  im = l84 * -3.0;
  if (im == 0.0) {
    re = ar / a_tmp;
    im = 0.0;
  } else if (ar == 0.0) {
    re = 0.0;
    im /= a_tmp;
  } else {
    re = ar / a_tmp;
    im /= a_tmp;
  }
  l115.re = (((re + l130_re) + l132_re) + l85) + l110.re;
  l115.im = (((im + l130_im) + l95) + l78) + l110.im;
  coder::internal::scalar::b_sqrt(&l115);
  l76 = l118_re * l12 - l118_im * l6;
  l3 = l118_re * l6 + l118_im * l12;
  l118_re = l76 * l115.re - l3 * l115.im;
  l118_im = l76 * l115.im + l3 * l115.re;
  if (l118_im == 0.0) {
    l6 = l118_re / 6.0;
    l12 = 0.0;
  } else if (l118_re == 0.0) {
    l6 = 0.0;
    l12 = l118_im / 6.0;
  } else {
    l6 = l118_re / 6.0;
    l12 = l118_im / 6.0;
  }
  ar = l86 * 3.0;
  im = l84 * 3.0;
  if (im == 0.0) {
    re = ar / a_tmp;
    im = 0.0;
  } else if (ar == 0.0) {
    re = 0.0;
    im /= a_tmp;
  } else {
    re = ar / a_tmp;
    im /= a_tmp;
  }
  l115.re = (((re + l130_re) + l132_re) + l85) + l110.re;
  l115.im = (((im + l130_im) + l95) + l78) + l110.im;
  coder::internal::scalar::b_sqrt(&l115);
  l118_re = l76 * l115.re - l3 * l115.im;
  l118_im = l76 * l115.im + l3 * l115.re;
  if (l118_im == 0.0) {
    l110.re = l118_re / 6.0;
    l110.im = 0.0;
  } else if (l118_re == 0.0) {
    l110.re = 0.0;
    l110.im = l118_im / 6.0;
  } else {
    l110.re = l118_re / 6.0;
    l110.im = l118_im / 6.0;
  }
  t3[0].re = 0.0;
  t3[0].im = 0.0;
  t3[1].re = l135_re - l110.re;
  t3[1].im = l74 - l110.im;
  t3[2].re = l135_re + l110.re;
  t3[2].im = l74 + l110.im;
  t3[3].re = -l135_re - l6;
  t3[3].im = -l74 - l12;
  t3[4].re = -l135_re + l6;
  t3[4].im = -l74 + l12;
  l110.re = l78_tmp * 2.0 - b_l78_tmp * 2.0;
  l12 = A_wayp * J_max;
  for (k = 0; k < 5; k++) {
    l3 = t3[k].re;
    l6 = t3[k].im;
    re = l3 * l3 - l6 * l6;
    l84 = l3 * l6;
    im = l84 + l84;
    ar = (((((l110.re + l8_tmp * re) - l2_tmp) + l5_tmp) - l81_tmp * l3 * 2.0) +
          l12 * l3 * 2.0) -
         l84_tmp * re;
    im = ((l8_tmp * im - l81_tmp * l6 * 2.0) + l12 * l6 * 2.0) - l84_tmp * im;
    l3 = l11_tmp * l3 * 2.0 - l84_tmp * l3 * 2.0;
    l6 = l11_tmp * l6 * 2.0 - l84_tmp * l6 * 2.0;
    if (l6 == 0.0) {
      if (im == 0.0) {
        re = ar / l3;
        im = 0.0;
      } else if (ar == 0.0) {
        re = 0.0;
        im /= l3;
      } else {
        re = ar / l3;
        im /= l3;
      }
    } else if (l3 == 0.0) {
      if (ar == 0.0) {
        re = im / l6;
        im = 0.0;
      } else if (im == 0.0) {
        re = 0.0;
        im = -(ar / l6);
      } else {
        re = im / l6;
        im = -(ar / l6);
      }
    } else {
      l86 = std::abs(l3);
      l76 = std::abs(l6);
      if (l86 > l76) {
        l76 = l6 / l3;
        l3 += l76 * l6;
        re = (ar + l76 * im) / l3;
        im = (im - l76 * ar) / l3;
      } else if (l76 == l86) {
        if (l3 > 0.0) {
          l76 = 0.5;
        } else {
          l76 = -0.5;
        }
        if (l6 > 0.0) {
          l3 = 0.5;
        } else {
          l3 = -0.5;
        }
        re = (ar * l76 + im * l3) / l86;
        im = (im * l76 - ar * l3) / l86;
      } else {
        l76 = l3 / l6;
        l3 = l6 + l76 * l3;
        re = (l76 * ar + im) / l3;
        im = (l76 * im - ar) / l3;
      }
    }
    l2[k].re = re;
    l2[k].im = im;
  }
  l110.re = A_init - A_wayp;
  for (k = 0; k < 5; k++) {
    ar = -((l110.re + J_min * t3[k].re) + J_max * l2[k].re);
    im = -(J_min * t3[k].im + J_max * l2[k].im);
    if (im == 0.0) {
      t[k].re = ar / J_max;
      t[k].im = 0.0;
    } else if (ar == 0.0) {
      t[k].re = 0.0;
      t[k].im = im / J_max;
    } else {
      t[k].re = ar / J_max;
      t[k].im = im / J_max;
    }
    t[k + 5].re = 0.0;
    t[k + 5].im = 0.0;
    t[k + 10] = t3[k];
    t[k + 15].re = 0.0;
    t[k + 15].im = 0.0;
    t[k + 20].re = 0.0;
    t[k + 20].im = 0.0;
    t[k + 25].re = 0.0;
    t[k + 25].im = 0.0;
    t[k + 30] = l2[k];
  }
}

// End of code generation (aceg_O_AVP.cpp)
